### ISSUE TITLE
node_modules/virgil-crypto/virgil_crypto_node.node: undefined symbol: _ZN2v87Isolate19CheckMemoryPressureEv

### ISSUE THREAD

**Issue by MihranSh99** [VirgilSecurity/virgil-crypto#26]
OS - Ubuntu
node version - 8.9.4
"virgil-crypto": "^3.2.1",
"virgil-sdk": "^5.2.2",
The error is very unstable, occurs time to time when restarting server

**Comment by vadimavdeev** [comment]
Could you please provide more details? What functions do you call when the error occurs? Do you run your server locally on your computer, in virtual machine or on some remote server? 
If you could provide a minimum sample project where the error can be reproduced, that would be great.

**Comment by MihranSh99** [comment]
The error occurs both local and remote servers. The problem is that I can't also reproduce the error constantly. It can work without problems for days, but one day may not work after restarting server or deploying the server, but sometimes it works after restarting.

**Comment by vadimavdeev** [comment]
After doing a quick web search, it seems that it's a bug in Node.js. See this [comment](https://github.com/Automattic/node-canvas/issues/1252#issuecomment-431713561). 
Can you upgrade Node.js to the latest LTS version (8.15.0 currently) and see if the problem persists?

**Comment by MihranSh99** [comment]
I tried but the same error

**Comment by vadimavdeev** [comment]
What operations do you perform with `virgil-crypto`? And how often do you perform them?

**Comment by MihranSh99** [comment]
Generating JWT token with request.
Not So often this operation is used

**Comment by vadimavdeev** [comment]
Could you provide a project that I can run to reproduce the issue (even unreliably)?

**Comment by MihranSh99** [comment]
Sorry but I can't.

**Comment by cardoso** [comment]
Closing this issue due to inactivity.

**Comment by hellogerard** [comment]
I noticed that `virgil-crypto` downloads pre-built binaries. Is there any way to compile this binary manually, locally in a build process?

**Comment by hellogerard** [comment]
So I managed to build this locally using `./utils/build.sh --target=nodejs-8.11.2`. Then I copy the binary over to my app and it is working. That's progress. But I'm not out of the woods yet because I'm not sure how to incorporate this into our CI process. Ideally, the `virgil-crypto-javascript` wrapper would compile this on install using `node-gyp`, which is what the `bcrypt` package does.

**Comment by SergeySeroshtan** [comment]
Hello @hellogerard.
Thanks for using VirgilSecurity products.
I'll take a look what can we do and give an answer in 24 hours.

**Comment by SergeySeroshtan** [comment]
Hello @hellogerard.

The building process of the `virgil-crypto` library and it's wrapper for `node` is complicated.
It involves a bunch of utilities: cmake, swig, git.
The previous attempt to write a `node-gyp` configuration file failed due to external dependencies.

The current `virgil-crypto` library version is LTS. And new features were added to the fully re-written library in C [virgil-crypto-c](https://github.com/VirgilSecurity/virgil-crypto-c).
Unfortunately the wrapper for `nodejs` is not ready yet, but it is currently in development and will be ready in early September.
The new `node` wrapper won't rely on `SWIG` or a specific `node` version, and it will use [N-API](https://nodejs.org/api/n-api.html#n_api_n_api).

**Comment by hellogerard** [comment]
@SergeySeroshtan Thank you for the prompt response. I'm still not quite sure what is going on but I have been able to work around this problem by building locally during CI. I understand the difficulties with `node-gyp` and look forward to the new `nodejs` version!

BTW, now that the server is working, I have run into another binary issue with `virgil-crypto` in the client app that I have posted in the other repo.

https://github.com/VirgilSecurity/virgil-crypto-javascript/issues/44
